src/modules/site-v2/base/utils/view_decorators.py

```diff
@@ -0,0 +1,224 @@
+      release = DatasetRelease.from_name(release_version, species_name=species.name)
+
+    # The `from_name` method always raises a NotFoundError if the name is not valid
+    except NotFoundError:
+      return abort(404)
+    
+    # If species name provided with underscore instead of dash, redirect to dashed version of URL
+    if species.get_slug() != species_name:
+      return redirect( url_for(request.endpoint, *args, species_name=species.get_slug(), release_version=release_version, **kwargs) )
+
+    # Pass the objects to the function
+    return f(*args, species=species, release=release, **kwargs)
+
+  return decorator
+
+
+
+def parse_job_id(pipeline_class: Type[JobPipeline], fetch=True, check_data_exists=True):
+  '''
+    Parse `report_id` string argument from the URL into a `JobPipeline` subclass object,
+    and pre-fetches the data and results if desired.
+
+    Aborts with `404` if report ID was invalid, or optionally if no input data could be fetched.
+
+    Arguments:
+      - `pipeline_class`: The `JobPipeline` subclass to use to lookup the report.
+      - `fetch`: If `True`, pre-fetch the input `data` and output `result`, and pass them to the wrapped function as keyword arguments.
+      - `check_data_exists`: If `True`, abort with `404` if the fetched input data is `None`.
+  '''
+
+  def wrapper(f):
+
+    @wraps(f)
+    def decorator(*args, report_id, **kwargs):
+
+      # Fetch requested phenotype report
+      # Ensures the report exists and the current user has permission to view it
+      try:
+        job = lookup_report(pipeline_class.get_kind(), report_id)
+
+      # If the report lookup request is invalid, show an error message
+      except ReportLookupError as ex:
+        flash(ex.msg, 'danger')
+        abort(ex.code)
+
+      # Optionally bail out here -- don't bother fetching data/results
+      if not fetch:
+        return f(*args, job=job, **kwargs)
+
+      # Try getting & parsing the report data file and results
+      # If result is None, job hasn't finished computing yet
+      try:
+        data, result = job.fetch()
+
+      # Error reading one of the report files
+      except (EmptyReportDataError, EmptyReportResultsError) as ex:
+        logger.error(f'Error fetching {pipeline_class.get_kind()} report {ex.id}: {ex.description}')
+        return abort(404, description = ex.description)
+
+      # Error with the submission data
+      # This should only be possible if a report was somehow created with invalid data,
+      # e.g. not enough traits in a Phenotype Analysis report
+      except DataValidationError as ex:
+        logger.error(f'Error fetching {pipeline_class.get_kind()} report {id}: {ex}')
+        flash(ex.msg, 'error')
+        return abort(400, description = ex.msg)
+
+      # General error
+      except Exception as ex:
+        logger.error(f'Error fetching {pipeline_class.get_kind()} report {id}: {ex}')
+        return abort(400, description = 'Something went wrong')
```

Comment:
prefer slightly more descriptive error messages that can help us troubleshoot issues down the road, especially to identify at what layer the issue occurred. If we don't want to disclose details on internals, we can come up with a ERROR_CODE set of enums (this would be something we can quickly identify on our end).

Reply:
Yep, that was my concern -- I'm hesitant to abort with an arbitrary Exception message, since it might (and probably should) contain internal details.  The internal ERROR_CODE idea sounds interesting... also wonder if we could attach an "error ID" to certain `abort`ed results that we track in the logs, so we can easily locate them in the GCP console?